fawltydeps/extract_declared_dependencies.py

```diff
@@ -288,6 +288,28 @@ def parse_pyproject_toml(path: Path) -> Iterator[DeclaredDependency]:
     else:
         logger.debug("%s does not contain [tool.poetry].", source)
 
+    if "dynamic" in parsed_contents.get("tool", {}).get("setuptools", {}).keys():
+        dynamic_section = parsed_contents["tool"]["setuptools"]["dynamic"]
+        dynamic_deps_files = dynamic_section.get("dependencies", {}).get("file", [])
+        dynamic_optional_deps_files = [
+            file_list
+            for optional_deps_item in [
+                optional_deps_section.get("file", {})
+                for optional_deps_section in dynamic_section.get(
+                    "optional-dependencies", {}
+                ).values()
+            ]
+            for file_list in optional_deps_item
+        ]
+        dynamic_files = dynamic_deps_files + dynamic_optional_deps_files
+        if isinstance(dynamic_files, list):
+            for req_file in dynamic_files:
+                req_file_path = Path(req_file)
+                if req_file_path.exists():
+                    yield from parse_requirements_txt(req_file_path)
+                else:
+                    logger.debug("%s does not exist. Skipping", req_file_path)
```

Comment:
Hmm.. If the file is not found, then we have a problem. I am tempted to error out in this case.
@jherland thoughts?

Reply:
Hmm. Maybe error out, or at least print a very visible warning/error message before continuing.

How do we currently handle other serious (but recoverable) errors in dep file parsing? E.g. a missing field in pyproject.toml or similar?

Reply:
It seems we're erroring out when a pyproject element is mistyped:
https://github.com/tweag/FawltyDeps/blob/86d07b28990bea0d5bc7a6f354176b9cac27ec3f/fawltydeps/extract_declared_dependencies.py#L259-L267

https://github.com/tweag/FawltyDeps/blob/86d07b28990bea0d5bc7a6f354176b9cac27ec3f/fawltydeps/extract_declared_dependencies.py#L207

https://github.com/tweag/FawltyDeps/blob/86d07b28990bea0d5bc7a6f354176b9cac27ec3f/fawltydeps/extract_declared_dependencies.py#L228

Reply:
The `logger.error(...)` on line 260 prints an error message, but the rest of the exception handler does not re-raise the exception, or otherwise abort execution. The surrounding loop keeps going, and we return normally at the end, AFAICS.

I suspect something similar can be done in this case:
- raise a `FileNotFoundError` (or maybe it is already raised by `parse_requirements_txt()`?) at the point where we try to open the requirements files
- catch this exception somewhere suitable, probably inside the `for` loop on line 306, so that we can keep trying to parse the other dynamic files?
- In the exception handler: log an error message, and then keep on running.

Reply:
>raise a FileNotFoundError (or maybe it is already raised by parse_requirements_txt()?) at the point where we try to open the requirements files

Oh that's a good point. If that's the case then our job is done for us.
Otherwise, both options 2 and 3 are ok. As long as we log the error very visibly, there's no need to error out.

Reply:
So I actually created a minimal pyproject.toml based on the ansible-lint pyproject.toml, and didn't create the actual requirements files. This is what I got when I tried building the package:
>[..]/python3.9/site-packages/setuptools/config/expand.py:134: SetuptoolsWarning: File '/home/nour/python-projects/pyproject-dynamic-deps/.config/requirements.in' cannot be found

Same for the other requirements files. The build concluded successfully and I only got these warnings with the verbose mode (`pip wheel . -w dist -v`). I still think we should show the error by default but maybe we can still continue the analysis.

Reply:
Thanks for digging into how setuptools handles this. Would be perfect to add that minimal project as a real_project (or sample_project) test?

Agree that the way to handle this is loud error message, but still continue.

Reply:
Addressed in https://github.com/tweag/FawltyDeps/pull/351/commits/4242069da6847018533f95b1a4af1ee86159a4a0